### PR TITLE
Add preprocessing support for knowledge base updates

### PR DESCRIPTION
**Pull request by dusvyat**
## Description

Introduce classes to handle document preprocessing before inserting data into the knowledge base. This includes configuring preprocessors, handling various input formats (files, SQL query results, web pages, raw rows), and integrating preprocessing into knowledge base operations.

The initial scope of this PR was for contextual chunk preprocessing, however this lays the structure for us to easily modify and add preprocessing steps to document ingestion to knowledge base.

image

Fixes ML-186

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] 📄 This change requires a documentation update

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Test Location: Specify the URL or path for testing.
 - [ ]   Verification Steps: Outline the steps or queries needed to validate the change. Include any data, configurations, or actions required to reproduce or see the new functionality.

## Additional Media:

- [ ] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [ ] I have appropriately commented on my code, especially in complex areas.
- [ ] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.

### CODE REVIEW

File: mindsdb/api/http/namespaces/knowledge_bases.py [hunk 1]
```diff
@@ -28,88 +20,6 @@
...
-    column_names = [c.get('alias', c.get('name')) for c in query_result.columns]
-    df_to_insert = pd.DataFrame.from_records(query_result.data, columns=column_names)
-    table.insert(df_to_insert)
-
-
-def _insert_rows_into_knowledge_base(rows: List[Dict], table: KnowledgeBaseTable):
-    if not rows:
-        return
-    df_to_insert = pd.DataFrame.from_records(rows)
-    table.insert(df_to_insert)
```
Comment (dusvyat) [inline mindsdb/api/http/namespaces/knowledge_bases.py:111]: moving this into a self contained module called preprocess.py

File: mindsdb/api/http/namespaces/knowledge_bases.py [hunk 2]
```diff
@@ -28,88 +20,6 @@
...
-
```
Comment (dusvyat) [inline mindsdb/api/http/namespaces/knowledge_bases.py:112]: moved to `preprocess.py`

File: mindsdb/integrations/handlers/chromadb_handler/chromadb_handler.py [hunk 3]
```diff
@@ -305,7 +305,7 @@ def select(
             TableField.ID.value: ids,
             TableField.CONTENT.value: documents,
             TableField.METADATA.value: metadatas,
-            TableField.EMBEDDINGS.value: embeddings.tolist(),
+            TableField.EMBEDDINGS.value: embeddings,
```
Comment (dusvyat) [inline mindsdb/integrations/handlers/chromadb_handler/chromadb_handler.py:308]: this was added in my an external contributor and it breaks the rag_pipeline as embeddings is already a List

**Comment by tino097** [comment]
Because some [errors](https://github.com/mindsdb/mindsdb/issues/10046#issuecomment-2439978023) that i got during manual testing of creating a KB using Elasticsearch, i have checkout this PR and test it again but im getting errors again. 

Not sure if im using this correctly but from the code and provided test i tried to create KB like this 
```
CREATE KNOWLEDGE_BASE kb_custom
USING
	    model = embedding_model,
        preprocessing = { 
            "type": "contextual",
            "provider": "mindsdb",
            "contextual_config": {
                    "chunk_size": 1000,
                    "chunk_overlap": 200,
                    "llm_model": 'gpt-4'
                }
        }
```
KB is created and looks like this 
image

However when i im trying to populate the data into KB im getting this errors:

```
2024-10-27 20:15:33,511            http ERROR    mindsdb.api.http.namespaces.sql: Error profiling query: 
Traceback (most recent call last):
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/http/namespaces/sql.py", line 49, in post
    result = mysql_proxy.process_query(query)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
    result = function(*args, **kwargs)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/mysql_proxy.py", line 497, in process_query
    executor.query_execute(sql)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
    result = function(*args, **kwargs)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/mysql_executor.py", line 102, in query_execute
    self.do_execute()
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
    result = function(*args, **kwargs)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/mysql_executor.py", line 134, in do_execute
    ret = self.command_executor.execute_command(self.query)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
    result = function(*args, **kwargs)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/command_executor.py", line 587, in execute_command
    SQLQuery(statement, session=self.session, execute=True, database=database_name)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 93, in __init__
    self.execute_query()
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 261, in execute_query
    raise e
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 256, in execute_query
    step_result = self.execute_step(step)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 315, in execute_step
    return handler(self, steps_data=steps_data).call(step)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/steps/insert_step.py", line 87, in call
    dn.create_table(
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/datahub/datanodes/project_datanode.py", line 176, in create_table
    kb_table = session.kb_controller.get_table(table_name, self.project.id)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 626, in get_table
    return KnowledgeBaseTable(kb, self.session)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 45, in __init__
    self.configure_preprocessing(kb.params['preprocessing'])
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 51, in configure_preprocessing
    self._preprocessor = PreprocessorFactory.create_preprocessor(preprocessing_config)
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/preprocess/models.py", line 148, in create_preprocessor
    return ContextualPreprocessor(config.contextual_config or ContextualConfig())
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/preprocess/models.py", line 104, in __init__
    self.llm = create_chat_model({"model_name": config.llm_model})
  File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/agents/langchain_agent.py", line 144, in create_chat_model
    if args["provider"] == "anthropic":
KeyError: 'provider'
```

Am I doing something wrong here ?

**Comment by dusvyat** [comment]
> Because some [errors](https://github.com/mindsdb/mindsdb/issues/10046#issuecomment-2439978023) that i got during manual testing of creating a KB using Elasticsearch, i have checkout this PR and test it again but im getting errors again.
> 
> Not sure if im using this correctly but from the code and provided test i tried to create KB like this
> 
> ```
> CREATE KNOWLEDGE_BASE kb_custom
> USING
> 	    model = embedding_model,
>         preprocessing = { 
>             "type": "contextual",
>             "provider": "mindsdb",
>             "contextual_config": {
>                     "chunk_size": 1000,
>                     "chunk_overlap": 200,
>                     "llm_model": 'gpt-4'
>                 }
>         }
> ```
> 
> KB is created and looks like this image
> 
> However when i im trying to populate the data into KB im getting this errors:
> 
> ```
> 2024-10-27 20:15:33,511            http ERROR    mindsdb.api.http.namespaces.sql: Error profiling query: 
> Traceback (most recent call last):
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/http/namespaces/sql.py", line 49, in post
>     result = mysql_proxy.process_query(query)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
>     result = function(*args, **kwargs)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/mysql_proxy.py", line 497, in process_query
>     executor.query_execute(sql)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
>     result = function(*args, **kwargs)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/mysql_executor.py", line 102, in query_execute
>     self.do_execute()
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
>     result = function(*args, **kwargs)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/mysql/mysql_proxy/executor/mysql_executor.py", line 134, in do_execute
>     ret = self.command_executor.execute_command(self.query)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/utilities/profiler/profiler.py", line 141, in wrapper
>     result = function(*args, **kwargs)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/command_executor.py", line 587, in execute_command
>     SQLQuery(statement, session=self.session, execute=True, database=database_name)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 93, in __init__
>     self.execute_query()
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 261, in execute_query
>     raise e
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 256, in execute_query
>     step_result = self.execute_step(step)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/sql_query.py", line 315, in execute_step
>     return handler(self, steps_data=steps_data).call(step)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/sql_query/steps/insert_step.py", line 87, in call
>     dn.create_table(
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/api/executor/datahub/datanodes/project_datanode.py", line 176, in create_table
>     kb_table = session.kb_controller.get_table(table_name, self.project.id)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 626, in get_table
>     return KnowledgeBaseTable(kb, self.session)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 45, in __init__
>     self.configure_preprocessing(kb.params['preprocessing'])
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/controller.py", line 51, in configure_preprocessing
>     self._preprocessor = PreprocessorFactory.create_preprocessor(preprocessing_config)
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/preprocess/models.py", line 148, in create_preprocessor
>     return ContextualPreprocessor(config.contextual_config or ContextualConfig())
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/knowledge_base/preprocess/models.py", line 104, in __init__
>     self.llm = create_chat_model({"model_name": config.llm_model})
>   File "/home/tino097/Projects/personal/mindsdb/mindsdb/./mindsdb/interfaces/agents/langchain_agent.py", line 144, in create_chat_model
>     if args["provider"] == "anthropic":
> KeyError: 'provider'
> ```
> 
> Am I doing something wrong here ?

@tino097 did you try already setting `provider ="openai"`?

**Comment by tino097** [comment]
> @tino097 do you have openai key set in env variables?
No, im afraid i haven't

**Comment by dusvyat** [comment]
> No, i didnt have set the API KEY as i dont have a openai account 
> 
> 

@tino097 I would recommend you to join the Mindsdb community slack and posting questions/problems in there 🙏🏼

**Pull request merged**